File: leetcode/python/62. 不同路径.py
#!/bin/bash python
"""
62. 不同路径

思路：
正常的二维DP
"""

class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        if m <= 0 or n <= 0:
            return 0
        dp = [[1] * n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0: continue
                top = dp[i - 1][j] if i >= 1 else 0
                lef = dp[i][j - 1] if j >= 1 else 0
                dp[i][j] = top + lef
        return dp[-1][-1]

File: leetcode/cpp/62. 不同路径.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        if (m <= 0 || n <= 0) {
            return 0;
        }
        vector<vector<int>> dp(m, vector<int>(n, 1));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0) continue;
                int top = (i >= 1 ? dp[i - 1][j] : 0);
                int lef = (j >= 1 ? dp[i][j - 1] : 0);
                dp[i][j] = top + lef;
            }
        }
        return dp[m - 1][n - 1];
    }
};